### PR TITLE
8303210: [linux, Windows] Make UseSystemMemoryBarrier available as product flag

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
I'd like to enable UseSystemMemoryBarrier by default on supported Operating Systems in order to improve performance of thread state transitions (I/O, JNI, foreign function calls, JIT compiler threads, etc.). See JBS issue for more details.
Unfortunately, the feature was not yet implemented on all platforms. I added the code, but need the platform maintainers to check if it can be used reliably (and ideally if the performance improves). It's easy to switch it off again in case of problems.

Update: Startup performance and some benchmarks on some platforms are impacted (see below). So, this PR no longer switches it on by default.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8303500](https://bugs.openjdk.org/browse/JDK-8303500) to be approved

### Issues
 * [JDK-8303210](https://bugs.openjdk.org/browse/JDK-8303210): [linux, Windows] Make UseSystemMemoryBarrier available as product flag
 * [JDK-8303500](https://bugs.openjdk.org/browse/JDK-8303500): Make UseSystemMemoryBarrier available as product flag (**CSR**)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [5fc7dbd6](https://git.openjdk.org/jdk/pull/12753/files/5fc7dbd6e06c37ff60beb78f5d48c3c0a057f8e2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12753/head:pull/12753` \
`$ git checkout pull/12753`

Update a local copy of the PR: \
`$ git checkout pull/12753` \
`$ git pull https://git.openjdk.org/jdk.git pull/12753/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12753`

View PR using the GUI difftool: \
`$ git pr show -t 12753`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12753.diff">https://git.openjdk.org/jdk/pull/12753.diff</a>

</details>

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
@bulasevic, @RealFYang, @offamitkumar, @shipilev: Can you please check your platforms and provide feedback?
Maybe the feature can also be implemented for MacOS. I haven't checked.

**Comment by dholmes-ora** [comment]
There  are two steps needed here:
1. An argument/justification that this move from being experimental to being product (even if diagnostic).
2. Enabling it by default.

**Comment by TheRealMDoerr** [comment]
> There are two steps needed here:
> 
> 1. An argument/justification that this move from being experimental to being product (even if diagnostic).
> 2. Enabling it by default.

Hi David, thanks for looking into this.
1. We want to support it in product JDKs. I think there has been enough testing to leave the experimental state. See https://github.com/openjdk/jdk/pull/12708#issuecomment-1442965746
2. We need it activated for customers to benefit from the performance improvement. In addition, I'd like to enable it as soon as possible in order to get additional testing before RDP1, also in EA builds. Switching it off again in case of any problems is very easy. Performance results were published here: https://github.com/openjdk/jdk/pull/10123#issuecomment-1235123012
In addition, performance measurements wrt. the foreign function interface (project Panama) have shown the memory barrier performance bottleneck.

File: src/hotspot/os/linux/systemMemoryBarrier_linux.cpp [hunk 1]
```diff
@@ -72,12 +72,12 @@ static int membarrier(int cmd, unsigned int flags, int cpu_id) {
 bool LinuxSystemMemoryBarrier::initialize() {
   int ret = membarrier(MEMBARRIER_CMD_QUERY, 0, 0);
   if (ret < 0) {
-    log_error(os)("MEMBARRIER_CMD_QUERY unsupported");
+    log_trace(os)("MEMBARRIER_CMD_QUERY unsupported");
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/os/linux/systemMemoryBarrier_linux.cpp:75]: Also log in the success case too, before returning.

File: src/hotspot/share/runtime/threads.cpp [hunk 2]
```diff
@@ -554,10 +555,11 @@ jint Threads::create_vm(JavaVMInitArgs* args, bool* canTryAgain) {
 
   if (UseSystemMemoryBarrier) {
     if (!SystemMemoryBarrier::initialize()) {
-      vm_shutdown_during_initialization("Failed to initialize the requested system memory barrier synchronization.");
-      return JNI_EINVAL;
+      if (!FLAG_IS_DEFAULT(UseSystemMemoryBarrier)) {
+        warning("UseSystemMemoryBarrier specified, but not supported on this OS.");
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/runtime/threads.cpp:559]: Also, with the flag checking this code now looks to be in the wrong place. Flag warnings typically belong is arguments.cpp.
Actually I think this code belongs in os::init() regardless.

**Review by dholmes-ora — COMMENTED** [review]
Thanks for the pointers to the other discussions. I think we need to see broad benchmarks runs on all affected platforms to ensure this isn't penalizing applications that don't actually need the fast native access.

It may have been better to split this PR in two, first adding support to the platforms missing it, and then changing the flag type and state.

A few comments below.

Thanks

**Comment by dholmes-ora** [comment]
I'm running a swag of our benchmarks on linux-x64 and linux-aarch64 with the flag simply set to true. Preliminary results show a big hit on startup benchmarks on both platforms!

**Comment by TheRealMDoerr** [comment]
> I'm running a swag of our benchmarks on linux-x64 and linux-aarch64 with the flag simply set to true. Preliminary results show a big hit on startup benchmarks on both platforms!

Interesting. I had only checked large benchmarks before. I don't know why startup is impacted. There were also concerns when many single thread handshakes are used. Maybe you're right and I should only make the feature available on all platforms for now (and keep it experimental and switched off).

**Comment by dholmes-ora** [comment]
> There were also concerns when many single thread handshakes are used.
Ah! The startup issues may be related to JDK-8300926. I'll need to apply that change and rerun them.

**Comment by dholmes-ora** [comment]
There is also a 7.9% regression on Aarch64 with the Renaissance-Scala-Kmeans benchmark.

**Comment by offamitkumar** [comment]
Build and tier1 test are okay with this patch, except one new failure (it could be just a random failure, unrelated to this patch)
compiler/jsr292/ContinuousCallSiteTargetChange.java 

But @TheRealMDoerr I'm not sure about how to perform regression, will catch up with Tyler for it and post the result.

**Comment by TheRealMDoerr** [comment]
> Build and tier1 test are okay with this patch, except one new failure (it could be just a random failure, unrelated to this patch) compiler/jsr292/ContinuousCallSiteTargetChange.java
> 
> But @TheRealMDoerr I'm not sure about how to perform regression, will catch up with Tyler for it and post the result.

Thanks for checking! "ContinuousCallSiteTargetChange.java" often fails. It's probably unrelated. I guess I'll change my plan and only make the feature available for now since there are some drawbacks which prevent us from enabling it by default on all platforms.

**Comment by TheRealMDoerr** [comment]
> There is also a 7.9% regression on Aarch64 with the Renaissance-Scala-Kmeans benchmark.
> 
> And 6% on Renaissance-ParMnemonics on Aarch64 (but a 7% improvement on x64 for this one).
> 
> This may have to be an opt-in potential optimisation, for which we would need a full-fledged product flag (not experimental or diagnostic).

Thanks a lot for measuring! It's very unfortunate that this nice OS feature performs poorly in some cases! So, I'll keep it off by default for now (hoping that there will be some improvement in the future).

**Comment by RealFYang** [comment]
bootcycle and tier1-3 test good on linux-riscv64 platform. I think it will be safer to explicitly include "runtime/globals.hpp":

```
diff --git a/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp b/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
index d0083f932df..117145ac137 100644
--- a/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
+++ b/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
@@ -44,6 +44,7 @@
 #include "prims/methodHandles.hpp"
 #include "runtime/continuation.hpp"
 #include "runtime/continuationEntry.inline.hpp"
+#include "runtime/globals.hpp"
 #include "runtime/jniHandles.hpp"
 #include "runtime/safepointMechanism.hpp"
 #include "runtime/sharedRuntime.hpp"
diff --git a/src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp b/src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp
index 3a259d4ba18..ff7b40c2695 100644
--- a/src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp
+++ b/src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp
@@ -45,6 +45,7 @@
 #include "runtime/arguments.hpp"
 #include "runtime/deoptimization.hpp"
 #include "runtime/frame.inline.hpp"
+#include "runtime/globals.hpp"
 #include "runtime/jniHandles.hpp"
 #include "runtime/sharedRuntime.hpp"
 #include "runtime/stubRoutines.hpp"
```

**Comment by TheRealMDoerr** [comment]
> bootcycle and tier1-3 test good on linux-riscv64 platform. I think it will be safer to explicitly include "runtime/globals.hpp":

Thanks for testing! Added the includes.

**Comment by TheRealMDoerr** [comment]
I have made it a regular product flag and keep it off by default. I'll need a CSR, now.

**Comment by robehn** [comment]
If there are performance regressions enabled it by default may not be the best thing for our users.

The issue in ContinuousCallSiteTargetChange.java is a bit worrying, since it concerns depots and transitions.

I don't mind this being a product flag.

**Review by robehn — APPROVED** [review]

**Comment by TheRealMDoerr** [comment]
Thanks for the review! Also for the CSR review.
@dholmes-ora: I think I've done all changes you had requested. The CSR is proposed.

File: src/hotspot/os/aix/os_aix.cpp [hunk 3]
```diff
@@ -2222,6 +2223,15 @@ static void set_page_size(size_t page_size) {
 
 // This is called _before_ the most of global arguments have been parsed.
 void os::init(void) {
+  if (UseSystemMemoryBarrier) {
+    if (!SystemMemoryBarrier::initialize()) {
+      if (!FLAG_IS_DEFAULT(UseSystemMemoryBarrier)) {
```
Comment (dholmes-ora) [inline src/hotspot/os/aix/os_aix.cpp:2228]: Have we processed this flag by this stage?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Approval in principal, but the initialization code needs fixing up.

> Note: The feature requires at least Linux 4.14. The flag needs to be off for older versions. 

In that case we should do a version check when trying to initialize, and at least warn if the OS is too old.

Thanks

File: src/hotspot/os/aix/os_aix.cpp [hunk 4]
```diff
@@ -2330,6 +2340,12 @@ void os::init(void) {
 
 // This is called _after_ the global arguments have been parsed.
 jint os::init_2(void) {
+  if (UseSystemMemoryBarrier) {
+    if (!FLAG_IS_DEFAULT(UseSystemMemoryBarrier)) {
+      warning("UseSystemMemoryBarrier specified, but not supported on this OS.");
+    }
+    FLAG_SET_ERGO(UseSystemMemoryBarrier, false);
+  }
```
[the remark this answers is not in the document]
Comment (TheRealMDoerr) [reply src/hotspot/os/aix/os_aix.cpp:2340]: Thanks! The one is os::init was the leftover. The flags are initialized in `init_2`.

File: src/hotspot/os/aix/os_aix.cpp (hunk 3 above)
Comment (TheRealMDoerr) [reply src/hotspot/os/aix/os_aix.cpp]: This is the leftover. See below.

**Comment by TheRealMDoerr** [comment]
@dholmes: Thanks for reviewing! I had moved the flag initialization to `os_init2()` because of your comment "Actually I think this code belongs in os::init() regardless.", but `os::init()` is too early. I actually agree with that somewhere in arguments.cpp would be nice, but I couldn't find a good place for it there.

**Comment by TheRealMDoerr** [comment]
> > Note: The feature requires at least Linux 4.14. The flag needs to be off for older versions.
> 
> In that case we should do a version check when trying to initialize, and at least warn if the OS is too old.

We already warn if the linux version is too old:
https://github.com/openjdk/jdk/pull/12753/files#diff-aeec57d804d56002f26a85359fc4ac8b48cfc249d57c656a30a63fc6bf3457adR4496
Would you like the message to mention Linux 4.14?

**Comment by dholmes-ora** [comment]
> > > Note: The feature requires at least Linux 4.14. The flag needs to be off for older versions.
> > 
> > 
> > In that case we should do a version check when trying to initialize, and at least warn if the OS is too old.
> 
> We already warn if the linux version is too old: https://github.com/openjdk/jdk/pull/12753/files#diff-aeec57d804d56002f26a85359fc4ac8b48cfc249d57c656a30a63fc6bf3457adR4496 Would you like the message to mention Linux 4.14?

I think I may have misunderstood that comment in JBS. I read it is as "you can turn this on in versions earlier than 4.14 but you shouldn't". If trying to turn it on in earlier versions just fails, then that is fine.

**Comment by dholmes-ora** [comment]
> @dholmes: Thanks for reviewing! I had moved the flag initialization to `os::init_2()` because of your comment "Actually I think this code belongs in os::init() regardless.", but `os::init()` is too early. I actually agree with that somewhere in arguments.cpp would be nice, but I couldn't find a good place for it there.

Yeah I think I mistakenly thought this would only affect Linux - sorry. Duplicating in ` os::init2` on all platforms is not good. As for placement in arguments.cpp ... I would suggest adding a little function and calling it after `os::init_container_support()`.

File: src/hotspot/os/aix/os_aix.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/os/aix/os_aix.cpp:2339]: But this code isn't callling `SystemMemoryBarrier::initialize` ??

Ah now I see - this is a hard-coded "flag not supported". But there is no need to do that as `SystemMemoryBarrier::initialize` tells you whether it is supported or not.

**Comment by TheRealMDoerr** [comment]
I already had a bad feeling replicating the code. I've put the flag configuration directly into systemMemoryBarrier.hpp (because there is no .cpp file). Is this ok?
The CSR is now provisional. Do we have to do anything or just wait?

**Review by dholmes-ora — APPROVED** [review]
Update looks good. Thanks.

The placement in a .hpp file is not ideal but lacking a cpp file ...

**Pull request closed without merging**